fms/utils/serialization.py

```diff
@@ -88,48 +107,117 @@ def get_adapted(
 from fms import models
 
 
+def get_ckp_format(model_path: Union[str, Path]) -> str:
+    """
+    Returns the checkpoint format of a model checkpoint. If format
+    is not recognized, assumes all files are regular pytorch
+    checkpoints and returns "pt"
+
+    Args:
+    model_path: the path to find the weights.
+    """
+    model_path = Path(os.path.expanduser(model_path))
+    if (
+        model_path.suffix == ".safetensors"
+        or len(sorted(model_path.glob("*.safetensors"))) > 0
+    ):
+        return "st"
+    if model_path.suffix == ".pth" or len(sorted(model_path.glob("*.pth"))) > 0:
+        return "pt"
+    if model_path.suffix == ".bin" or len(sorted(model_path.glob("*.bin"))) > 0:
+        return "hf"
+    return "pt_all"
+
+
+def get_safetensors_item(key, file: Path, device: torch.device) -> torch.Tensor:
+    from safetensors import safe_open  # type: ignore[import-untyped]
+
+    with torch.no_grad():
+        with safe_open(
+            file, framework="pt", device=str(device)
+        ) as model_weights:  # type: ignore[attr-defined]
+            return model_weights.get_tensor(key)
+
+
+class LazySafetensorsDict(collections.UserDict):
+    def set_lazy_tensor(self, key, file, device):
+        super().__setitem__(key, lambda: get_safetensors_item(key, file, device))
+
+    def __getitem__(self, key):
+        lazy_tensor = super().__getitem__(key)
+        if callable(lazy_tensor):
+            lazy_tensor = lazy_tensor()
+            super().__setitem__(key, lazy_tensor)
+        return lazy_tensor
+
+
 def load_state_dict(
     model_path: Union[str, Path],
+    *,
+    checkpoint_format: Optional[str] = None,
     distributed_strategy: Optional[str] = None,
     checkpoint_sharding: Optional[str] = None,
     initial_device: torch.device = torch.device("cpu"),
     rank: int = 0,
     world_size: int = 1,
-) -> Mapping[str, Any]:
+) -> MutableMapping[str, Any]:
```

Comment:
what required this to be mutable?

Reply:
I am deleting all the weights in the state_dict once they are copied to the model so we don't allocate them twice. This requires making the state_dict Mutable

Reply:
Which reminds me I should mention that the state dict will be consumed in the `load_state_dict_into_model` function